File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;
class Solution {
    public long solution(int n, int[] times) {
        long answer = 0;
        Arrays.sort(times);
        long left = times[0];
        long right = times[times.length-1] * (long) n;
        
        while (left <= right) {
            long mid = (left + right) / 2;
            long people = 0;
            for (int time : times) {
                people += mid / time;
                if (people >= n) {
                    break;
                }
            }
            
            if (people >= n) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return answer;
    }
}

File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    answer = 0
    left = 0
    right = max(times) * n;
    while (left <= right):
        mid = (left + right) // 2;
        people = 0
        for time in times:
            people += mid // time
            if people >= n:
                break
                
        # n명을 넘겨서 심사 한거면 시간이 너무 많은 상황
        # 딱 n명을 심사했더라도 시간은 남을 수 있음
        if people >= n:
            answer = mid
            right = mid - 1
        else:
            left = mid + 1
        
    return answer